### ISSUE TITLE
MySQL NO_ZERO_DATE prevents table creation

### ISSUE THREAD

**Issue by jsfry** [woocommerce/action-scheduler#587]
**Edit to add:** This is a followup issue to #517 where the `_actions` table doesn't get created due to the `NO_ZERO_DATE` being set in the MySQL configuration in version 5.7.

WordPress version 5.4.2
Current theme: BizPlus Pro (version 1.0.3)
Current plugin: Action Scheduler (version 3.1.6)
PHP version 7.4.5
mysql 8.0.20
WC 3.9.2 & 4.3.0
AS 3.1.6

This started a couple days ago when my website threw a wsod and I was notified by WP recovery mode that it was due to the following:
Error Details
=============
An error of type E_ERROR was caused in line 44 of the file /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php. Error message: Uncaught RuntimeException: Error saving action: Error saving action: Table 'wordpress.wp_actionscheduler_actions' doesn't exist in /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php:44
Stack trace:
#0 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(242): ActionScheduler_DBStoreMigrator->save_action()
#1 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_ActionFactory.php(177): ActionScheduler_HybridStore->save_action()
#2 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_ActionFactory.php(84): ActionScheduler_ActionFactory->store()
#3 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/functions.php(36): ActionScheduler_ActionFactory->single()
#4 /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/migration/Scheduler.php(89): as_s

I have tried to resolve this via your instructions on the page: https://github.com/woocommerce/woocommerce/issues/25895
But With either WC 3.9.2 or WC 4.3.0 as soon as I activate the AS plugin I get the WSOD immediately due to no tables being present - see below for error on that from my apache logfile.  Not sure where to go from here - any advice pls?

------ apache log -----
[Mon Jul 13 12:22:35.230090 2020] [php7:notice] [pid 32029] [client 127.0.0.1:32838] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='in-progress' ORDER BY scheduled_date_gmt DESC LIMIT 1 made by require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->is_migration_scheduled, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-cron.php?doing_wp_cron=1594639355.0161380767822265625000
[Mon Jul 13 12:22:35.231102 2020] [php7:notice] [pid 32029] [client 127.0.0.1:32838] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='pending' ORDER BY scheduled_date_gmt ASC LIMIT 1 made by require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->is_migration_scheduled, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-cron.php?doing_wp_cron=1594639355.0161380767822265625000
[Mon Jul 13 12:22:35.232064 2020] [php7:notice] [pid 32029] [client 127.0.0.1:32838] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='in-progress' ORDER BY scheduled_date_gmt DESC LIMIT 1 made by require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->schedule_migration, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-cron.php?doing_wp_cron=1594639355.0161380767822265625000
[Mon Jul 13 12:22:35.232879 2020] [php7:notice] [pid 32029] [client 127.0.0.1:32838] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='pending' ORDER BY scheduled_date_gmt ASC LIMIT 1 made by require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->schedule_migration, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-cron.php?doing_wp_cron=1594639355.0161380767822265625000
[Mon Jul 13 12:22:35.240124 2020] [php7:notice] [pid 32029] [client 127.0.0.1:32838] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SHOW FULL COLUMNS FROM `wp_actionscheduler_actions` made by require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->schedule_migration, as_schedule_single_action, ActionScheduler_ActionFactory->single, ActionScheduler_ActionFactory->store, ActionScheduler_HybridStore->save_action, ActionScheduler_DBStoreMigrator->save_action, ActionScheduler_DBStore->save_action, referer: https://birthlight.com/wp-cron.php?doing_wp_cron=1594639355.0161380767822265625000
[Mon Jul 13 12:22:35.240242 2020] [php7:error] [pid 32029] [client 127.0.0.1:32838] PHP Fatal error:  Uncaught RuntimeException: Error saving action: Error saving action: Table 'wordpress.wp_actionscheduler_actions' doesn't exist in /var/www/html/wp-content/plugins/action-scheduler-master/classes/migration/ActionScheduler_DBStoreMigrator.php:44\nStack trace:\n#0 /var/www/html/wp-content/plugins/action-scheduler-master/classes/data-stores/ActionScheduler_HybridStore.php(242): ActionScheduler_DBStoreMigrator->save_action()\n#1 /var/www/html/wp-content/plugins/action-scheduler-master/classes/ActionScheduler_ActionFactory.php(177): ActionScheduler_HybridStore->save_action()\n#2 /var/www/html/wp-content/plugins/action-scheduler-master/classes/ActionScheduler_ActionFactory.php(84): ActionScheduler_ActionFactory->store()\n#3 /var/www/html/wp-content/plugins/action-scheduler-master/functions.php(36): ActionScheduler_ActionFactory->single()\n#4 /var/www/html/wp-content/plugins/action-scheduler-master/classes/migration/Scheduler.php(89): as_schedule_single_action()\n#5 /var/www/html/wp-content/plugins/action-scheduler-master/ in /var/www/html/wp-content/plugins/action-scheduler-master/classes/migration/ActionScheduler_DBStoreMigrator.php on line 44, referer: https://birthlight.com/wp-cron.php?doing_wp_cron=1594639355.0161380767822265625000
[Mon Jul 13 12:22:35.282791 2020] [php7:notice] [pid 32025] [client 31.48.69.185:42694] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='in-progress' ORDER BY scheduled_date_gmt DESC LIMIT 1 made by require_once('wp-admin/admin.php'), require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->is_migration_scheduled, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-admin/plugins.php?plugin_status=all&paged=1&s
[Mon Jul 13 12:22:35.283715 2020] [php7:notice] [pid 32025] [client 31.48.69.185:42694] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='pending' ORDER BY scheduled_date_gmt ASC LIMIT 1 made by require_once('wp-admin/admin.php'), require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->is_migration_scheduled, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-admin/plugins.php?plugin_status=all&paged=1&s
[Mon Jul 13 12:22:35.284650 2020] [php7:notice] [pid 32025] [client 31.48.69.185:42694] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='in-progress' ORDER BY scheduled_date_gmt DESC LIMIT 1 made by require_once('wp-admin/admin.php'), require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->schedule_migration, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-admin/plugins.php?plugin_status=all&paged=1&s
[Mon Jul 13 12:22:35.285452 2020] [php7:notice] [pid 32025] [client 31.48.69.185:42694] WordPress database error Table 'wordpress.wp_actionscheduler_actions' doesn't exist for query SELECT a.action_id FROM wp_actionscheduler_actions a WHERE a.hook='action_scheduler/migration_hook' AND a.status='pending' ORDER BY scheduled_date_gmt ASC LIMIT 1 made by require_once('wp-admin/admin.php'), require_once('wp-load.php'), require_once('wp-config.php'), require_once('wp-settings.php'), do_action('wp_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Action_Scheduler\\Migration\\Controller->schedule_migration, Action_Scheduler\\Migration\\Scheduler->schedule_migration, as_next_scheduled_action, ActionScheduler_HybridStore->find_action, ActionScheduler_DBStore->find_action, referer: https://birthlight.com/wp-admin/plugins.php?plugin_status=all&paged=1&s

**Comment by rrennick** [comment]
Hi @jsfry, Could you run the following SQL command in your database admin tool (eg. phpMyAdmin)

```
CREATE TABLE wp_actionscheduler_actions (
				        action_id bigint(20) unsigned NOT NULL auto_increment,
				        hook varchar(191) NOT NULL,
				        status varchar(20) NOT NULL,
				        scheduled_date_gmt datetime NOT NULL default '0000-00-00 00:00:00',
				        scheduled_date_local datetime NOT NULL default '0000-00-00 00:00:00',
				        args varchar(191),
				        schedule longtext,
				        group_id bigint(20) unsigned NOT NULL default '0',
				        attempts int(11) NOT NULL default '0',
				        last_attempt_gmt datetime NOT NULL default '0000-00-00 00:00:00',
				        last_attempt_local datetime NOT NULL default '0000-00-00 00:00:00',
				        claim_id bigint(20) unsigned NOT NULL default '0',
				        extended_args varchar(8000) DEFAULT NULL,
				        PRIMARY KEY  (action_id),
				        KEY hook (hook(191)),
				        KEY status (status),
				        KEY scheduled_date_gmt (scheduled_date_gmt),
				        KEY args (args(191)),
				        KEY group_id (group_id),
				        KEY last_attempt_gmt (last_attempt_gmt),
				        KEY claim_id (claim_id)
				        ) ${charset_collate}
```

Note that `${charset_collate}` needs to be replaced with the charset & collate values from you other WP tables. That's the statement that AS uses to create the table so if there is an error message please let us know what it is.

**Comment by jsfry** [comment]
Thanks @rrennick for getting back to me - much appreciated.  I've done what you asked and here's the error which was returned (did it on the command line - I suppose that's the same as phpmyadmin):

ERROR 1067 (42000): Invalid default value for 'scheduled_date_gmt'

Any further help would be much appreciated in order to get my woocommerce back up and running.

Thanks,
Jeff

**Comment by jsfry** [comment]
Btw just so you know and so you can check it (I'm no expert at mysql) ,  I sourced this file on the command line of mysql after saying "use <db>" where <db> is the name of my wordpress db.

CREATE TABLE wp_actionscheduler_actions (
				        action_id bigint(20) unsigned NOT NULL auto_increment,
				        hook varchar(191) NOT NULL,
				        status varchar(20) NOT NULL,
				        scheduled_date_gmt datetime NOT NULL default '0000-00-00 00:00:00',
				        scheduled_date_local datetime NOT NULL default '0000-00-00 00:00:00',
				        args varchar(191),
				        schedule longtext,
				        group_id bigint(20) unsigned NOT NULL default '0',
				        attempts int(11) NOT NULL default '0',
				        last_attempt_gmt datetime NOT NULL default '0000-00-00 00:00:00',
				        last_attempt_local datetime NOT NULL default '0000-00-00 00:00:00',
				        claim_id bigint(20) unsigned NOT NULL default '0',
				        extended_args varchar(8000) DEFAULT NULL,
				        PRIMARY KEY  (action_id),
				        KEY hook (hook(191)),
				        KEY status (status),
				        KEY scheduled_date_gmt (scheduled_date_gmt),
				        KEY args (args(191)),
				        KEY group_id (group_id),
				        KEY last_attempt_gmt (last_attempt_gmt),
				        KEY claim_id (claim_id)
				        ) CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci;

**Comment by rrennick** [comment]
@jsfry Thanks for the follow up. You should be able to create the table with

```
CREATE TABLE `wp_actionscheduler_actions` (
  `action_id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `hook` varchar(191) COLLATE utf8mb4_unicode_520_ci NOT NULL,
  `status` varchar(20) COLLATE utf8mb4_unicode_520_ci NOT NULL,
  `scheduled_date_gmt` datetime DEFAULT NULL,
  `scheduled_date_local` datetime DEFAULT NULL,
  `args` varchar(191) COLLATE utf8mb4_unicode_520_ci DEFAULT NULL,
  `schedule` longtext COLLATE utf8mb4_unicode_520_ci,
  `group_id` bigint(20) unsigned NOT NULL DEFAULT '0',
  `attempts` int(11) NOT NULL DEFAULT '0',
  `last_attempt_gmt` datetime DEFAULT NULL,
  `last_attempt_local` datetime DEFAULT NULL,
  `claim_id` bigint(20) unsigned NOT NULL DEFAULT '0',
  `extended_args` varchar(8000) COLLATE utf8mb4_unicode_520_ci DEFAULT NULL,
  PRIMARY KEY (`action_id`),
  KEY `hook` (`hook`),
  KEY `status` (`status`),
  KEY `scheduled_date_gmt` (`scheduled_date_gmt`),
  KEY `args` (`args`),
  KEY `group_id` (`group_id`),
  KEY `last_attempt_gmt` (`last_attempt_gmt`),
  KEY `claim_id` (`claim_id`)
) CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci;
```

I'm going to update the description of the issue to add an explanation for the error. Can you let us know if the other three action scheduler tables get created okay once this one is in place?

**Comment by jsfry** [comment]
Thanks for further advice.  The response from mysql was: Query OK, 0 rows affected, 4 warnings (0.14 sec).  

I will proceed to run the plugin now to see if I can get the latest WC running and get back to you.

**Comment by jsfry** [comment]
The Action Scheduler plugin installed fine.  As I mentioned the other 3 (claims, groups, and logs) were already there before the fix above went in for the actions table.  So, next I clicked on Tools -> Scheduled Actions  and it did its thing.  Then I updated WC to the latest version and it started to update  the WC database.  The only slightly concerning thing in it all was that 20/21 of the database update items finished with 1 still pending (ran overnight) but at the same time it said "database updated successfully" so I'm not sure what that one pending item is and if it is significant enough to worry about.  I have a screenshot if you think it's significant.

**Comment by jonasrafael** [comment]
thanks @rrennick works like a charm!

**Comment by rrennick** [comment]
@jonasrafael Thanks for letting us know :)

**Comment by rrennick** [comment]
> in the edit which you did above you mentioned in the title caption a mysql version of 5.7 but I'm running mysql 8.0.20, not 5.7

As far as I know the NO_ZERO_DATE issue is specific to MySQL 5.7. If going to the dashboard screen creates the table then it was something else that prevented the table from being created.

**Comment by danieliser** [comment]
@rrennick We are seeing a lot of these with our user base (Popup Maker).

I can't reasonably ask all of these users to run custom queries like this, so what are the actual options for fixing this in the code base correctly?

I compared your code that worked for him:
- you added specific character encoding to the text based fields.
- removed `NOT NULL` from the datetime fields.
- replaced the `DEFAULT '0000-00-00 00:00:00'` with `DEFAULT NULL`
- added backticks (which we can't do reliably with dbDelta from my understanding)

Are these viable changes to make within the AS table declaration itself?

Also if we patched with those changes in place, how would we reliably tell AS to reinstall them. Is there an option key we purge?

**Comment by rrennick** [comment]
> I compared your code that worked for him:

The code I provided above is a copy from the previous comment so that it matched the charset/collation. Those can be changed to match any of the charset/collation combinations supported by WP and the current MySQL install.

> added backticks

Those aren't necessary for any of the Action Scheduler schema.

> I can't reasonably ask all of these users to run custom queries like this, so what are the actual options for fixing this in the code base correctly?

Changing the schema is easy. The barrier at this point is transitioning existing tables to a declaration that is supported by dbDelta on MySQL configured with this issue. 

An interim fix would be to create a small plugin that checks for the table & if it doesn't exist create it using the `DEFAULT NULL` on the date fields.

**Comment by danieliser** [comment]
@rrennick Other than potential for empty dates to be saved, is there some regression this could cause? I'm not seeing it, but I'm not deep into all the use cases.

I'm assuming that since the data doesn't actually get changed, there really isn't a migration neccessary. No field types were changed, only the default values and whether allowed to be null or not.

We run many DB migrations but I've never had to do one for such simple changes.

That said I did not see anything that indicated those tables were individually versioned, so I'm thinking that is what your referring to needing to implement (checking for table version and updating when needed).

**Comment by rrennick** [comment]
> Other than potential for empty dates to be saved, is there some regression this could cause?

Action Scheduler doesn't rely on the default value for the date fields. The fields were modelled on the `wp_posts` table in WP. WP turns off the two configuration settings that cause this issue when connecting to the DB. It's not clear why this does not happen with the core table but does with the AS table.

**Comment by danieliser** [comment]
@rrennick - Agree with you there, so I spent some time digging into WP core, and unless your tables are being created outside of the existing global WPDB connection I can't see how this is possible UNLESS their MySQL is not allowing WP to change modes: https://github.com/WordPress/WordPress/blob/d936f2c9591893597ff3fee145225251c5482e8d/wp-includes/wp-db.php#L864-L872

Now that said, something that has occurred to me is that with the rarity of this issue there may be something specific at play.

Observations:
- Nobody is reporting NO_ZERO_DATE mysql errors. These only came from users who were trying to manually run queries directly on DB, thus not bypassing incomp modes.
- You won't see any errors at all IF show_errors is false or suppress_errors is true. Only way you should see errors is if WP_DEBUG && WP_DEBUG_DISPLAY  are both true.
- dbDelta suppressing errors may actually allow a query to finish.

My thinking is that this is only occurring to users with WP_DEBUG and WP_DEBUG_DISPLAY which would explain why so few see it.

But that should also mean that there is potentially a silent error occurring on initial install that is getting logged but not displayed. Without that original error it would be hard to track this down further.

In conclusion here, I don't think the new title and description of this issue is accurate, and it may indicate many of the other errors linked to this issue are more related than originally thought.

**Comment by rrennick** [comment]
>  I can't see how this is possible UNLESS their MySQL is not allowing WP to change modes

Right. I believe that is the issue.

> Nobody is reporting NO_ZERO_DATE mysql errors.

I've been able to reproduce the error locally. Adding and removing that setting determine whether AS successfully creates the table or not.

**Comment by danieliser** [comment]
@rrennick - Correct I could duplicate it when I go into MySQL directly and run the table create queries. I ran into this exact same issue when writing a recent update to our analytics. But running the same queries through WPDB didn't throw the errors. I didn't get issues with NO_ZERO_DATE through $wpdb.

That said in my case it failed with generic errors in "your query after ..., check the docs for your version" etc. I see no evidence this query ever gets attempted.

**Comment by rrennick** [comment]
> I see no evidence this query ever gets attempted.

Thanks, I'll have a look at that.

**Comment by danieliser** [comment]
@rrennick To be clear, I'm confident the code is well written, but my gut says that with no errors for the failed creation of the table, those are either being eaten by WP error logging some how while letting others through, or the entire query simply never gets processed for some reason on these sites.

**Comment by danieliser** [comment]
That said I did just randomly stumble upon this potential workaround, which if we can confirm somehow that it does work for these users would be easy enough:  https://dev.to/anraiki/comment/i3gb

**Comment by rrennick** [comment]
@danieliser Nice find, thank you.

**Comment by rrennick** [comment]
There has to be some other configuration parameter required to reproduce this:

```
mysql> SELECT @@GLOBAL.sql_mode;
+----------------------------------+
| @@GLOBAL.sql_mode                |
+----------------------------------+
| STRICT_TRANS_TABLES,NO_ZERO_DATE |
+----------------------------------+
1 row in set (0.00 sec)

mysql> drop table wp_actionscheduler_actions;
Query OK, 0 rows affected (0.00 sec)

mysql> show tables like '%act%';
+----------------------------------------+
| Tables_in_ron (%act%)                  |
+----------------------------------------+
| wp_actionscheduler_claims              |
| wp_actionscheduler_groups              |
| wp_actionscheduler_logs                |
| wp_as_actions                          |
| wp_wc_admin_note_actions               |
| wp_woocommerce_product_key_activations |
+----------------------------------------+
6 rows in set (0.00 sec)

-- Refresh Tools -> Scheduled Actions

mysql> show tables like '%act%';
+----------------------------------------+
| Tables_in_ron (%act%)                  |
+----------------------------------------+
| wp_actionscheduler_actions             |
| wp_actionscheduler_claims              |
| wp_actionscheduler_groups              |
| wp_actionscheduler_logs                |
| wp_as_actions                          |
| wp_wc_admin_note_actions               |
| wp_woocommerce_product_key_activations |
+----------------------------------------+
7 rows in set (0.00 sec)
```

**Comment by danieliser** [comment]
I will say I NEVER saw issues with `NO_ZERO_DATE` errors when my queries were run through `$wpdb`.

What I saw was more generalized errors from the position before or after the `default '0000-` section of the query. Only when I tried to run the same queries in MySQL did I see those specific issues which led me to simply change to another field type in our case. Not sure that is viable here though, so `DEFAULT NULL` maybe the safest option long term.

Granted I also ran into issues because of new reserved keywords used in MariaDB such as we tried naming a field `returning` but MariaDB has that as a reserved keyword in the latest versions.

But WP always gave generic MySQL errors, and never spit out anything close to resembling that issue in its error messages. So very well could be some other issue that is getting swallowed up.

**Comment by efel80** [comment]
Hi Daniel and Rrennik. I just run into the problem described here (and in other posts, but i think this is the right one).

 I'm not an expert or a programmer, just a prosumer of wordpress/woocommerce from years on live sites used by thousands.

I'm trying to  understand how to resolve this problem and help who encounter this issue.
Hope this can help you to understand. We own some ecommerce on live production that works from 5 years + and moved all our sites from different ecommerce systems to woocommerce in the last years to have a better centralized management and specialize on it.
We moved with caution from version to version of wordpress and woocommerce in the last years, without any problem. Waiting for maturity of updates and fixes. Putting on sites quickly only security updates and bugfix or in case the update is useful to get better performances. 

Now the issue: 
We had to move a woocommerce site from a server to another. 
Old server (DEDICATED): DEBIAN 9 - MARIADB (Mysql 5.5.5) -10.1.26-MariaDB-0+deb9u1 - PHP 7.4.9

This is the status page: 
`WC Version: 4.4.1
REST API Version: ✔ 1.0.15
WC Blocks Version: ✔ 3.1.0
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 1.4.0
Log Directory Writable: ✔
WP Version: 5.5.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: it_IT
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.25 (Debian)
PHP Version: 7.4.9
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2u

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.1.26-MariaDB-0+deb9u1
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Il tuo server non ha la classe SoapClient attiva - alcuni plugin gateway che usano SOAP potrebbero non funzionare correttamente.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 4.4.1
WC Database Prefix: wpls_
Dimensione totale database: 532.60MB
Dimensione dati database: 464.73MB
Dimensione indice database: 67.87MB
wpls_woocommerce_sessions: Dati: 7.74MB + indice: 0.20MB + motore MyISAM
wpls_woocommerce_api_keys: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_attribute_taxonomies: Dati: 0.00MB + indice: 0.01MB + motore MyISAM
wpls_woocommerce_downloadable_product_permissions: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_order_items: Dati: 0.06MB + indice: 0.03MB + motore MyISAM
wpls_woocommerce_order_itemmeta: Dati: 0.30MB + indice: 0.18MB + motore MyISAM
wpls_woocommerce_tax_rates: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_tax_rate_locations: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_shipping_zones: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_shipping_zone_locations: Dati: 0.00MB + indice: 0.01MB + motore MyISAM
wpls_woocommerce_shipping_zone_methods: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_payment_tokens: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_payment_tokenmeta: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woocommerce_log: Dati: 0.02MB + indice: 0.02MB + motore InnoDB
wpls_actionscheduler_actions: Dati: 19.55MB + indice: 11.61MB + motore InnoDB
wpls_actionscheduler_claims: Dati: 0.41MB + indice: 0.23MB + motore InnoDB
wpls_actionscheduler_groups: Dati: 0.02MB + indice: 0.02MB + motore InnoDB
wpls_actionscheduler_logs: Dati: 3.52MB + indice: 3.03MB + motore InnoDB
wpls_amazon_accounts: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_amazon_btg: Dati: 0.02MB + indice: 0.06MB + motore InnoDB
wpls_amazon_categories: Dati: 0.02MB + indice: 0.03MB + motore InnoDB
wpls_amazon_feeds: Dati: 0.03MB + indice: 0.00MB + motore InnoDB
wpls_amazon_feed_templates: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_amazon_feed_tpl_data: Dati: 0.02MB + indice: 0.08MB + motore InnoDB
wpls_amazon_feed_tpl_values: Dati: 0.02MB + indice: 0.05MB + motore InnoDB
wpls_amazon_jobs: Dati: 0.08MB + indice: 0.00MB + motore InnoDB
wpls_amazon_listings: Dati: 24.56MB + indice: 2.19MB + motore InnoDB
wpls_amazon_log: Dati: 0.02MB + indice: 0.05MB + motore InnoDB
wpls_amazon_markets: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_amazon_orders: Dati: 0.08MB + indice: 0.05MB + motore InnoDB
wpls_amazon_payment: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_amazon_profiles: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_amazon_reports: Dati: 171.52MB + indice: 0.00MB + motore InnoDB
wpls_amazon_shipping: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_amazon_stock_log: Dati: 0.44MB + indice: 0.00MB + motore InnoDB
wpls_commentmeta: Dati: 0.02MB + indice: 0.02MB + motore MyISAM
wpls_comments: Dati: 0.29MB + indice: 0.12MB + motore MyISAM
wpls_iwp_backup_status: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_iwp_processed_iterator: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_links: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_mg_amazonawsapiaccounts: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_mg_amazonglobal: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_mg_amazonmwsapiaccounts: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_mg_amazonstores: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_mg_amazon_regionlist: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_mg_amazon_reportlist: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_mg_amazon_reporttype: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_ms_snippets: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_nextend2_image_storage: Dati: 0.02MB + indice: 0.02MB + motore InnoDB
wpls_nextend2_section_storage: Dati: 0.02MB + indice: 0.03MB + motore InnoDB
wpls_nextend2_smartslider3_generators: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_nextend2_smartslider3_sliders: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_nextend2_smartslider3_sliders_xref: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_nextend2_smartslider3_slides: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_options: Dati: 13.63MB + indice: 1.57MB + motore MyISAM
wpls_postmeta: Dati: 137.96MB + indice: 26.59MB + motore MyISAM
wpls_posts: Dati: 66.97MB + indice: 6.64MB + motore MyISAM
wpls_ps_exclude: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_ps_postmeta: Dati: 0.53MB + indice: 0.33MB + motore MyISAM
wpls_ps_posts: Dati: 1.36MB + indice: 0.28MB + motore MyISAM
wpls_ps_product_sku: Dati: 0.40MB + indice: 0.28MB + motore MyISAM
wpls_revslider_css: Dati: 0.09MB + indice: 0.00MB + motore MyISAM
wpls_revslider_layer_animations: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_revslider_navigations: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_revslider_sliders: Dati: 0.01MB + indice: 0.00MB + motore MyISAM
wpls_revslider_slides: Dati: 0.05MB + indice: 0.00MB + motore MyISAM
wpls_revslider_static_slides: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_rsgiftvoucher: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_rspointexpiry: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_rsrecordpoints: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_rs_expiredpoints_email: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_rs_templates_email: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_snippets: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_sumo_reward_encashing_submitted_data: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_sumo_reward_send_point_submitted_data: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_termmeta: Dati: 0.72MB + indice: 0.50MB + motore MyISAM
wpls_terms: Dati: 0.66MB + indice: 0.86MB + motore MyISAM
wpls_term_relationships: Dati: 2.62MB + indice: 4.64MB + motore MyISAM
wpls_term_taxonomy: Dati: 0.64MB + indice: 0.59MB + motore MyISAM
wpls_tmsuserscontacts: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_tmsusersprofiles: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_usermeta: Dati: 0.06MB + indice: 0.04MB + motore MyISAM
wpls_users: Dati: 0.00MB + indice: 0.01MB + motore MyISAM
wpls_wcpdf_credit_note_number: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wcpdf_invoice_number: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wcpdf_packing_slip_number: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wcpdf_proforma_number: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wcpdf_receipt_number: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wc_admin_notes: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wc_admin_note_actions: Dati: 0.02MB + indice: 0.02MB + motore InnoDB
wpls_wc_category_lookup: Dati: 0.05MB + indice: 0.00MB + motore InnoDB
wpls_wc_customer_lookup: Dati: 0.06MB + indice: 0.03MB + motore InnoDB
wpls_wc_download_log: Dati: 0.02MB + indice: 0.03MB + motore InnoDB
wpls_wc_groupbuy_log: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wc_order_coupon_lookup: Dati: 0.02MB + indice: 0.03MB + motore InnoDB
wpls_wc_order_product_lookup: Dati: 0.08MB + indice: 0.06MB + motore InnoDB
wpls_wc_order_stats: Dati: 0.08MB + indice: 0.05MB + motore InnoDB
wpls_wc_order_tax_lookup: Dati: 0.02MB + indice: 0.03MB + motore InnoDB
wpls_wc_product_meta_lookup: Dati: 3.52MB + indice: 6.00MB + motore InnoDB
wpls_wc_reserved_stock: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wc_tax_rate_classes: Dati: 0.02MB + indice: 0.02MB + motore InnoDB
wpls_wc_webhooks: Dati: 0.02MB + indice: 0.02MB + motore InnoDB
wpls_wfBadLeechers: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfBlockedIPLog: Dati: 0.07MB + indice: 0.05MB + motore MyISAM
wpls_wfBlocks: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfBlocksAdv: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfConfig: Dati: 0.01MB + indice: 0.01MB + motore MyISAM
wpls_wfCrawlers: Dati: 0.01MB + indice: 0.00MB + motore MyISAM
wpls_wfFileMods: Dati: 0.89MB + indice: 0.24MB + motore MyISAM
wpls_wfHits: Dati: 0.38MB + indice: 0.09MB + motore MyISAM
wpls_wfHoover: Dati: 0.44MB + indice: 0.09MB + motore MyISAM
wpls_wfIssues: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfKnownFileList: Dati: 0.58MB + indice: 0.08MB + motore MyISAM
wpls_wfLeechers: Dati: 0.11MB + indice: 0.13MB + motore MyISAM
wpls_wfLockedOut: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfLocs: Dati: 0.01MB + indice: 0.00MB + motore MyISAM
wpls_wfLogins: Dati: 0.03MB + indice: 0.01MB + motore MyISAM
wpls_wfNet404s: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfNotifications: Dati: 0.01MB + indice: 0.00MB + motore MyISAM
wpls_wfReverseCache: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfScanners: Dati: 0.01MB + indice: 0.01MB + motore MyISAM
wpls_wfSNIPCache: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfStatus: Dati: 0.07MB + indice: 0.03MB + motore MyISAM
wpls_wfThrottleLog: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_wfVulnScanners: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_woobookstore_cache: Dati: 0.02MB + indice: 0.02MB + motore InnoDB
wpls_woocommerce_gpf_google_taxonomy: Dati: 1.02MB + indice: 0.00MB + motore MyISAM
wpls_wpforms_tasks_meta: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wpmailsmtp_tasks_meta: Dati: 0.02MB + indice: 0.00MB + motore InnoDB
wpls_wpmelon_advbedit_temp: Dati: 0.48MB + indice: 0.42MB + motore MyISAM
wpls_yith_wcwl: Dati: 0.00MB + indice: 0.00MB + motore MyISAM
wpls_yith_wcwl_lists: Dati: 0.00MB + indice: 0.01MB + motore MyISAM
wpls_yoast_seo_links: Dati: 0.06MB + indice: 0.02MB + motore InnoDB
wpls_yoast_seo_meta: Dati: 1.52MB + indice: 0.00MB + motore InnoDB

### Post Type Counts ###

amn_wpforms-lite: 1
attachment: 40883
authorbook: 31
custom_css: 1
jetpack_migration: 2
jp_img_sitemap: 33
jp_img_sitemap_index: 3
jp_sitemap: 3
jp_sitemap_master: 3
ml-slide: 3
ml-slider: 1
nav_menu_item: 73
page: 21
popup: 1
popup_theme: 7
post: 6
product: 30048
publisher: 84
retailers: 1
revision: 248
shop_coupon: 3
shop_order: 355
shop_order_refund: 27
store: 1
team: 4
wpforms: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (35) ###

InfiniteWP - Client: by Revmakx – 1.9.4.7
VaultPress: by Automattic – 2.1.4
WPini Update Toolkit: by WPini – 5.0
Akismet Anti-Spam: by Automattic – 4.1.6
AMP: by Collaboratori del progetto AMP – 2.0.1
Categories Images: by Muhammad El Zahlan – 3.0.1
Classic Editor: by Contributori WordPress – 1.6
Code Snippets: by Code Snippets Pro – 2.14.0
Coming Soon Page, Under Construction & Maintenance Mode by SeedProd: by SeedProd – 5.1.3
EU Cookie Law: by Alex Moss
Marco Milesi – 3.1.4

Facebook for WooCommerce: by Facebook – 2.0.1 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce Group Buy: by wpgenie – 1.1.11 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
Jetpack by WordPress.com: by Automattic – 8.9
Regenerate Thumbnails: by Alex Mills (Viper007Bond) – 3.1.3
Simple Image Sizes: by Rahe – 3.2.1
SUMO Discounts: by Fantastic Plugins – 4.0 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
UpdraftPlus - Backup/Ripristino: by UpdraftPlus.Com
DavidAnderson – 1.16.29

WooCommerce Smart COD: by woosmartcod.com – 1.6.1
WooCommerce Fattureincloud: by Woofatture – 1.9.5
Woocommerce Advanced Bulk Edit: by George Iron – 3.9.2 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce Amazon and eBay Integration: by Codisto – 1.3.52 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
Checkout Manager for WooCommerce: by QuadLayers – 5.2.7
WooCommerce Custom Field Product Search: by Alvaro Neto – 1.0.0 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce Expand Tabs: by SilkyPress – 1.17
WooCommerce Google Ads Conversion Tracking: by Wolf+Bär Agency – 1.6.9
WooCommerce Grid / List toggle: by jameskoster – 1.2.1 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce PDF Invoices Italian Add-on: by laboratorio d'Avanguardia – 0.7.0.17
WooCommerce PDF Invoices & Packing Slips: by Ewout Fernhout – 2.6.1
WooCommerce PDF Invoices & Packing Slips Professional: by Ewout Fernhout – 2.6.6
WooCommerce Product Archive Customiser: by jameskoster – 1.0.5 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce: by Automattic – 4.4.1 (aggiornamento alla versione 4.5.1 disponibile)
WP-Lister Lite for Amazon: by WP Lab – 1.7
WP Mail SMTP: by WPForms – 2.3.1
WP Overnight Sidekick: by Jeremiah Prummer
Ewout Fernhout
Michael Kluver – 2.2.2

YITH WooCommerce Wishlist: by YITH – 3.0.13

### Inactive Plugins (7) ###

Finale Lite - Sales Countdown Timer & Discount for WooCommerce: by XLPlugins – 2.13.0
Smart Manager For WooCommerce – Stock Management, Bulk Edit & more...: by StoreApps – 4.4.3
SUMO Reward Points: by Fantastic Plugins – 21.1 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce Admin: by WooCommerce – 1.5.0 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce Bookstore: by WPini – 3.0 – Versione installata non testata con versione attiva di WooCommerce 4.4.1
WooCommerce Product Feed: by WebAppick – 3.7.15
WPForms Lite: by WPForms – 1.6.2.3

### Must Use Plugins (1) ###

InfiniteWP - Client Loader: by Revmakx –

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: book (book)
external (external)
groupbuy (groupbuy)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔

### WC Pages ###

Shop base: #5 - /negozio/
Carrello: #6 - /carrello2/
Pagamento: #7 - /cassa/
Il mio account: #8 - /mio-account/
Termini e condizioni: #52482 - /termini-e-condizioni-duso/

### Theme ###

Name: Bookshop
Version: 1.0.18
Author URL: https://woocommerce.com
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 2.6.0 – 2.7.0 è disponibile
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Completato: 1.288
Oldest: 2020-08-22 13:25:16 +0200
Newest: 2020-09-13 11:01:39 +0200

Operazione non riuscita: 45.436
Oldest: 2020-04-30 14:14:25 +0200
Newest: 2020-08-02 22:49:20 +0200

In attesa: 2
Oldest: 2020-09-13 12:01:39 +0200
Newest: 2020-09-14 02:57:43 +0200
`
I'm trying to move an ecommerce site from a debian server to another with more resources. 
More Ram, better processor, new OS: MARIADB (Mysql 5.5.5) -10.1.26-MariaDB-0+deb9u1 - PHP 7.4.9

The problem is that if i move the site the site this is slow. 
And when i update to Woocommerce 4.5.1 i got the error.

`An error of type E_ERROR was caused in line 44 of the file /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php. Error message: Uncaught RuntimeException: Error saving action: Error saving action: Table 'wordpress.wp_actionscheduler_actions' doesn't exist in /var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php:44`

I tried to move the site manually and also using the backup from updraft.
I got jetpack, but i think (it need tiem) is unuseful at this point. 

All the configuration are the same or highers in term of post max size, time limit for operations, upload etc.

I have the popup maker plugin.
But really i didn't see solutions. Tried to disable plugins. 
Now i come back to old server and i have the idea to make there the upgrade and then pass to new server.
The problem is that if the problem persist we are "dead". 

As i said is the first time in years of using Wordpress + Woocommerce i got this type of devasting error.

Thank you
Efel

**Comment by efel80** [comment]
I see but don't understand sorry if is a common problem fixed or not.

**Comment by olnerpsm** [comment]
Hi everyone, I am getting the same issue with my WooCommerce plugin. Anything past 3.9.3 just resolves in a fatal error.

WordPress version 5.6
Current theme: Flatsome Child (version 3.13)
PHP version 7.3.25

Error Details
=============
An error of type E_ERROR was caused in line 44 of the file /public_html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php. Error message: Uncaught RuntimeException: Error saving action: Error saving action: Table 'lctadmin_database.wp_actionscheduler_actions' doesn't exist in /public_html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/migration/ActionScheduler_DBStoreMigrator.php:44
Stack trace:
#0 /public_html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/data-stores/ActionScheduler_HybridStore.php(225): ActionScheduler_DBStoreMigrator->save_action(Object(ActionScheduler_Action), NULL)
#1 /public_html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_ActionFactory.php(177): ActionScheduler_HybridStore->save_action(Object(ActionScheduler_Action))
#2 /public_html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_ActionFactory.php(84): ActionScheduler_ActionFactory->store(Object(ActionScheduler_Action))
#3 /public_html/wp-content/plugins/woocommerce/packages/action-scheduler/funct


I've tried using the SQL command at the top of this thread but Im getting an error:

Error
Static analysis:

3 errors were found during analysis.

Unexpected character. (near "$" at position 1161)
Unexpected character. (near "{" at position 1162)
Unexpected character. (near "}" at position 1189)
SQL query:

CREATE TABLE wp_actionscheduler_actions ( action_id bigint(20) unsigned NOT NULL auto_increment, hook varchar(191) NOT NULL, status varchar(20) NOT NULL, scheduled_date_gmt datetime NOT NULL default '0000-00-00 00:00:00', scheduled_date_local datetime NOT NULL default '0000-00-00 00:00:00', args varchar(191), schedule longtext, group_id bigint(20) unsigned NOT NULL default '0', attempts int(11) NOT NULL default '0', last_attempt_gmt datetime NOT NULL default '0000-00-00 00:00:00', last_attempt_local datetime NOT NULL default '0000-00-00 00:00:00', claim_id bigint(20) unsigned NOT NULL default '0', extended_args varchar(8000) DEFAULT NULL, PRIMARY KEY (action_id), KEY hook (hook(191)), KEY status (status), KEY scheduled_date_gmt (scheduled_date_gmt), KEY args (args(191)), KEY group_id (group_id), KEY last_attempt_gmt (last_attempt_gmt), KEY claim_id (claim_id) ) ${utf8mb4_utf8mb4_unicode_ci}

MySQL said: Documentation

#1064 - You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '${utf8mb4_utf8mb4_unicode_ci}' at line 23


Can anyone please help? I'm really struggling and have done for a while now to try and get this sorted. Thank you

**Comment by rrennick** [comment]
> Note that ${charset_collate} needs to be replaced with the charset & collate values from you other WP tables.

@olnerpsm `${some-string}` is a PHP interpolated string which PHP replaces. `${utf8mb4_utf8mb4_unicode_ci}` should be `utf8mb4_unicode_ci`.

**Comment by danieliser** [comment]
@rrennick Has there been any movement on this issue in general? Has a determination been made as to the ultimate cause, potential fixes etc? Or is the only workaround to have tables generated manually somehow?

**Comment by rrennick** [comment]
@danieliser The PR to address the issue is waiting for review. Because it involves a schema change it planned to go into a major release.

**Comment by danieliser** [comment]
@rrennick - Hmm, good news and bad news I guess. What is the timeline for v4 then given that seems to be the target?

**Comment by rrennick** [comment]
> What is the timeline for v4

It will be 3.2.0. We hope to have it included in WooCommerce 5.0.

**Comment by danieliser** [comment]
@rrennick Ahh, wouldn't that be a `minor` version then, or were you referring to WooCommerce 5 as the `major` version change?

**Comment by rrennick** [comment]
Action Scheduler follows the WordPress versioning scheme (X.X.0 is a major version) vs WooCommerce.

**Comment by vedanshujain** [comment]
Closing this as per https://github.com/woocommerce/action-scheduler/pull/725. @barryhughes feel free to re-open if that is not the case.

**Comment by barryhughes** [comment]
Good catch, thank you: that seems fine to me (remaining related scope is captured in #653 which remains open).
